### DISCUSSION TITLE
Decorator taking a (covariant?) callable and returning it again

### DISCUSSION THREAD

**Discussion by Bluenix2** [microsoft/pyright#2501] [Q&A] [answered]
Hello I have code such as the following:
```python
from typing import Callable, TypeVar

C = TypeVar('C', bound='Callable[[str], None]')


def do_nothing(func: C) -> C:
    return func


class StrSubclass(str):
    pass


@do_nothing
def test(arg: StrSubclass) -> None:
    pass
```
But this causes the following error:
```
Argument of type "(arg: StrSubclass) -> None" cannot be assigned to parameter "func" of type "C@do_nothing" in function "do_nothing"
  Type "(str) -> None" cannot be assigned to type "(arg: StrSubclass) -> None"
```
How would I type this code?

I've tried both `covariant` and `contravariant` but neither of them allow being used as a return type *and* parameter type.

I could reproduce this with [MyPy here](https://mypy-play.net/?mypy=latest&python=3.10&gist=8756beffabfd9776e65f58990a9d4469) so it doesn't seem like a bug with Pyright.

**Comment by erictraut** [comment, marked as answer]
Parameters within callables are implicitly contravariant, which is why you're seeing the error.

Think about it this way. The function `test` requires that the caller pass a `StrSubclass` instance as the first argument. But `C` is bound to a callback that accepts a `str`. If the decorator `do_nothing` were to call this callback with a `str` argument, it wouldn't satisfy the type requirements of `test`.

The workaround is to either change the bound type in `C` to accept a `StrSubclass` or change `test` to accept a `str` argument.

**Reply by Bluenix2** [reply]
Okay I think I understand why it is contravariant but I think I would need to make it covariant in some way..

During runtime I actually do some introspection and instantiate the specific type of the annotation. I am mostly looking to statically enforce that this is an event subclass.

That said I have since realized that I would like to allow the user to pass a callable with more arguments (but they would need to have a default), so I will just rely on my runtime checks.
Thanks.
